tests/unit_tests/device/cuda/container_cuda_kernel.hpp

```diff
@@ -9,24 +9,83 @@
 
 // Detray Core include(s)
 #include "detray/core/detail/multi_store.hpp"
+#include "detray/core/detail/single_store.hpp"
+#include "detray/core/detail/tuple_container.hpp"
 
 // Vecmem include(s)
 #include "vecmem/containers/device_vector.hpp"
 
-// Thrust include(s)
-#include <thrust/tuple.h>
-
 namespace detray {
 
-using host_store_type =
+// Single store test
+/// @{
+using single_store_t = single_store<double, vecmem::vector>;
+using single_store_dev_t = single_store<double, vecmem::device_vector>;
+/// @}
+
+// Tuple container test
+/// @{
+using tuple_cont_t = detail::tuple_container<dtuple, vecmem::vector<int>,
+                                             vecmem::vector<double>>;
+using tuple_cont_dev_t =
+    detail::tuple_container<dtuple, vecmem::device_vector<int>,
+                            vecmem::device_vector<double>>;
+/// @}
+
+// Regular multi store test (uses vectors as containers in every tuple element)
+/// @{
+using reg_multi_store_t =
     regular_multi_store<int, empty_context, dtuple, vecmem::vector, std::size_t,
                         float, double>;
-
-using device_store_type =
+using reg_multi_store_dev_t =
     regular_multi_store<int, empty_context, dtuple, vecmem::device_vector,
                         std::size_t, float, double>;
+/// @}
+
+/// Multi store test
+/// @{
+
+/// Test type that holds vecemem members and forces a hierarchical view/buffer
+/// treatment
+template <template <typename...> class vector_t = dvector>
+struct test {
+
+    using view_type = dmulti_view<dvector_view<int>, dvector_view<double>>;
+    using buffer_type =
+        dmulti_buffer<dvector_buffer<int>, dvector_buffer<double>>;
+
+    DETRAY_HOST test(vecmem::memory_resource* mr) : first(mr), second(mr) {}
+
+    template <typename view_t,
+              std::enable_if_t<detail::is_device_view_v<view_t>, bool> = true>
+    DETRAY_HOST_DEVICE test(view_t v)
+        : first(detail::get<0>(v.m_view)), second(detail::get<1>(v.m_view)) {}
+
+    DETRAY_HOST view_type get_data() {
+        return view_type{vecmem::get_data(first), vecmem::get_data(second)};
+    }
```

Comment:
I'm not sure I understand this, but I thought this was supposed not to be necessary to create as a member of each class we want to use, but that instead by defining view_type and buffer_type we could then just call detray::get_data(test) which would implicitly do this.

Reply:
Well, ```get_data(test)``` will just call ```test.get_data()``` because it otherwise has no access to the potentially private members of ```test``` and also cannot know what names they have. Only the class itself knows how to handle its members and how to construct its potentially completely custom view type (we don't have any completely custom view types, yet). The rest of detray should just be concerned with wrapping those views and passing them on to the class constructor. Of course a class with a totally custom view type would also have to provide the corresponding overloads of ```get_data``` and ```get_buffer``` for them, so that they can be called in the recursion.
But this is just at the idea stage, I have never tested this, so maybe it makes no sense at all :D